src/lib/provable/gadgets/blake2b.ts

```diff
@@ -0,0 +1,283 @@
+// https://datatracker.ietf.org/doc/html/rfc7693.html
+import { UInt64, UInt8 } from '../int.js';
+import { FlexibleBytes } from '../bytes.js';
+import { Bytes } from '../wrapped-classes.js';
+import { Gadgets } from './gadgets.js';
+import { assert } from '../../util/errors.js';
+import { Provable } from '../provable.js';
+
+export { BLAKE2B };
+
+/**
+ * IV[0..7]  Initialization Vector (constant).
+ * SIGMA[0..9]  Message word permutations (constant).
+ * p[0..7]  Parameter block (defines hash and key sizes).
+ * m[0..15]  Sixteen words of a single message block.
+ * h[0..7]  Internal state of the hash.
+ * d[0..dd-1]  Padded input blocks.  Each has "bb" bytes.
+ * t  Message byte offset at the end of the current block.
+ * f  Flag indicating the last block.
+ *
+ * All mathematical operations are on 64-bit words in BLAKE2b.
+ *
+ * Byte (octet) streams are interpreted as words in little-endian order,
+ * with the least-significant byte first.
+ */
+
+type State = {
+  h: UInt64[];
+  t: UInt64[];
```

Comment:
no, `t` is set to `[Uint64.zero, Uint64.zero]` at initialisation. `t[0]` is incremented by `buflen` during every `update()` and `final()`.  However, t[1] is only used when input size is larger than 2^64 bytes which is unlikely. We can brake up the `t `into two `Uint64`s instead of an array, but would it improve efficiency?